aip/auth/4118.md

```diff
@@ -0,0 +1,144 @@
+
+In order for workload credentials to be properly used, the auth libraries
+**must** support the automatic switching of the service endpoint to its mTLS
+counterpart.
+
+### Using Workload Credentials
+
+Users **should** configure [ADC][2] to use workload credentials via the
+certificate configuration gcloud metadata file. Workload credentials can be
+added as a **"cert_info"** type as follows:
+
+```json
+{
+  “version”: 1
+  "cert_info": {
+    "workload": {
+      "cert_path": "path/to/cert/file"
+      "key_path": "path/to/key/file"
+      “workload_identity_provider”: “...”
+      “service_account_email”: “...”
+    },
+    "keychain": {
+      ...
+    },
+    "pkcs11": {
+      ...
+    },
+    "windows": {
+      ...
+    },
+  },
+  "libs": {
+   ...
+  }
+}
+```
+
+For Linux and macOS platforms, the above metadata file is located in the
+well-known gcloud config directory at
+**"~/.config/gcloud/certificate_config.json"**. Note that the default location
+of this file can be changed using the `GOOGLE_API_CERTIFICATE_CONFIG`
+environment variable.
+
+The following lists the fields of the **"workload"** certificate info type that
+are relevant to workload credentials:
+
+  - **"cert_path"**: The specified value will be used as the full path to locate
+    the workload certificate file. This file **must** contain a PEM-encoded
+    X.509 certificate chain (ordered from leaf to root) where the leaf
+    certificate is a valid X.509 SVID. The chain __may__ consist of only the
+    leaf certificate.
+  - **"key_path"**: The specified value will be used as the full path to locate
+    the workload private key file. This file must contain a PEM-formatted
+    private key associated with the X.509 certificate specified by
+    **“cert_path”**.
+
+The description of the **"workload_identity_provider"** and
+**"service_account_email"** fields can be found in [mTLS Token Binding][5].
+
+To enable mutual authentication to Google APIs using workload credentials, the
+**"workload"** section and its **"cert_path"** and **"key_path"** values must be
+present in the **"~/.config/gcloud/certificate_config.json"** configuration
+file.
+
+### Expected Behavior
+
+Support for mTLS authentication to Google APIs using workload credentials
+**must** give priority to user mTLS endpoint override via client options. The
+auth libraries **must** follow the steps below:
+
+  - Locate the workload certificate and private key files using the above
+    config file. If one of these files is not present, mTLS using workload
+    credentials may be disabled. The auth libraries **must** check that the
+    public and private keys in the certificate and key files match before
+    passing them to the TLS library.
+      - If the certificate and private key files are loaded in memory (as
+        opposed to being read from disk for every mTLS connection), the auth
+        libraries **must** periodically reload them (at least every 10 minutes)
```

Comment:
May also consider checking for expiry time on the cert and reloading as necessary.

Reply:
Updated the text. Thanks for the suggestion!